// @sassoftware/va-report-components/dist/esm/chunk0.mjs
export { H as HeaderName, d as appReadyEvent, a as axios, c as configuredLanguage, h as createAxiosInstance, e as eventBus, j as getConfiguration, b as getLanguage, g as getTextLocalizer, k as initConfiguration, i as initI18next, l as loadBundle, o as onLanguageChange, s as setLanguage, f as signoutEvent, w as waitForInit } from "./chunk1.mjs";

export { AxiosHeaders, formToJSON, getAdapter, isAxiosError, isCancel, toFormData } from "axios";

import "react";

import "react-dom/client";

import "tslib";

import "rxjs";

import "rxjs/operators";

import "i18next";

import "react-i18next";

import "@babel/runtime/helpers/extends";

import "react-dom";

import "react-focus-lock";

import "classnames";

import "framer-motion";

import "polished";

import "reselect";

import "./i18n/nova_i18n.mjs";

import "react-popper";

import "@popperjs/core";

import "memoize-one";

import "handlebars";

import "zod";

import "use-memo-one";

import "@tanstack/react-query";

import "react-cropper";

import "redux-observable";

import "redux";

import "@redux-devtools/extension";

import "react-redux";
