// @sassoftware/va-report-components/dist/esm/index.mjs
/*! License information is available at ../../LICENSE.txt and ../../LICENSE-3RD-PARTY.txt */
export { a7 as SASReportElement, a8 as SASReportObjectElement, a9 as SASReportPageElement, aa as connectToServer, ab as registerDataDrivenContent, ac as setFormatterLocale, ad as setHonorLocalFormatSettings, ae as setLoadingTheme, af as setLocale, ag as setUseHighContrastReportTheme } from "./chunk1.mjs";

import "redux";

import "react";

import "polished";

import "react-i18next";

import "i18next";

import "classnames";

import "axios";

import "@babel/runtime/helpers/extends";

import "react-redux";

import "react-dom/client";

import "tslib";

import "rxjs";

import "rxjs/operators";

import "react-dom";

import "react-focus-lock";

import "framer-motion";

import "reselect";

import "./i18n/nova_i18n.mjs";

import "react-popper";

import "@popperjs/core";

import "memoize-one";

import "handlebars";

import "zod";

import "use-memo-one";

import "@tanstack/react-query";

import "react-cropper";

import "redux-observable";

import "@redux-devtools/extension";
